### ISSUE TITLE
Add Input Validation Support to WinUI (UWP and Win32 Desktop) with INotifyDataErrorInfo

### ISSUE THREAD

**Issue by thomasclaudiushuber** [microsoft/microsoft-ui-xaml#179]
# Proposal: Add Input Validation Support with INotifyDataErrorInfo 

## Summary
Add Input Validation Support with INotifyDataErrorInfo that is supported by x:Bind and Binding. Both Markup Extensions should get a new ValidatesOnNotifyDataErrors property that is - like in WPF's Binding - true by default. 

## Rationale
Currently UWP does not have any Input Validation built into the platform. But every line of business app requires input validation. It's one of the most important features for a proper enterprise app. There's an entry on uservoice that says this feature will come with WinUI, but I haven't seen anything yet: https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/13052589-uwp-input-validation

**Comment by jevansaks** [comment]
@LucasHaines how does this relate to the input validation features you've been looking into?

**Comment by LucasHaines** [comment]
@jevansaks This is directly related to the Input Validation work i'm outlining.

**Comment by thomasclaudiushuber** [comment]
Great. If you've any preview stuff to share, I'm happy to try it and provide feedback and thoughts.

**Comment by mdtauk** [comment]
![image](https://user-images.githubusercontent.com/7389110/51212818-5b651380-1911-11e9-8b1a-11654b1a434a.png)

This is the kind of validation mentioned during Build 2018

`
<TextBox
     x:Name="UserName" Header="User Name:" 
     Text="{x:Bind ViewModel.Person.UserName, 
	UpdateSourceTrigger=PropertyChanged, 
	Mode=TwoWay}" />

<PasswordBox 
     x:Name="Password" Header="Password:" 
     Password="{x:Bind ViewModel.Person.Password, 
                 UpdateSourceTrigger=LostFocus, 
                 Mode=TwoWay}"/>
`

**Comment by stevenbrix** [comment]
@thomasclaudiushuber how important do you think making this work as part of {Binding} is? I only ask because the technical challenges there are non-trivial, as well as it not being able to work downlevel. Our original idea was to just support x:Bind, which will only require markup compiler changes, and is capable of working downlevel.

Also, we were only planning on supporting the INotifyDataErrorInfo and DataAnnotations paradigm. We weren't intending on adding something analogous to Binding.ValidationRules, and so we didn't feel there was sufficent need for a ValidatesOnNotifyDataErrors.

We'd love to get your feedback on the idea

**Comment by thomasclaudiushuber** [comment]
@stevenbrix 

Short answer: Yes, what you think and what you plan sounds great. Just x:Bind is fine, and just INotifyDataErrorInfo is fine too.


Long:
### Just x:Bind? 👍 
In all WPF-LOB cases I can think of there was always a kind of ViewModel that was known at compile-time, no duck typing, so x:Bind is fine. I wrote Binding in this issue too as I thought you can support it to get the same syntax as in WPF. But this is more "nice to have" than super important. And as Binding and x:Bind are two completely different things behind the scenes and the runtime stuff with Binding is non-trivial like you mentioned, forget about Binding. I think having it just for x:Bind is totally fine, it will work for the use cases I can think of. And from all the conference talks I did on UWP and x:Bind, I can tell you that x:Bind is one of the most favorite UWP features of all XAML developers and I told all of them "Prefer x:Bind over Binding wherever you can". :) Getting intellisense, perf, step-into-code, and compile-time errors makes x:Bind the new default on UWP, so having the validation support just there is fine and a good decision imo. 

### Just INotifyDataErrorInfo and DataAnnotations paradigm? 👍 
Just supporting INotifyDataErrorInfo and DataAnnotations paradigm sounds good to me. WPF doesn't pick up Data Annotations automatically, you need to hook them up manually in your INotifyDataErrorInfo implementation by using the Validator class. You mean this when you say "DataAnnotations paradigm", right? Or do you plan built-in DataAnnotation support that allows a dev to just put a Data Annotation on a property and this just works? Like in ASP.NET Core MVC? While that would be great, I think it's not necessary. Having the classes ValidationContext, Validator and ValidationResult and the other classes from System.ComponentModel.DataAnnotations is sufficient imo.

### Add Binding.ValidationRules? Noooooo ;-)
No, definitely don't add that. INotifyDataErrorInfo is enough and its the best. I think I never used ValidationRules anymore after IDataErrorInfo support was added to WPF (If I remember correctly this was in .NET Framework 3.5). And when INotifyDataErrorInfo was added in .NET Framework 4.5, my customers and I have used that interface in all projects for input validation. Supporting just this interface is fine.

### Add ValidatesOnNotifyDataErrors? No, but...  
Yes, you don't have to add this one. But it has nothing to do with the Binding.ValidationRules. This property is directly related to the INotifyDataErrorInfo interface. On WPF's Binding Markup Extension, ValidatesOnNotifyDataErrors is by default true, which means the Binding picks up the implemented INotifyDataErrorInfo validation if the bound object implements that interface. If you set the ValidatesOnNotifyDataErrors property on the Binding markup extension to false, then the Binding Markup Extension won't pick up the INotifyDataErrorInfo implementation for the validation. So, maybe this isn't too hard to implement with x:Bind, but maybe we don't need it. So, no, don't do it. That's ok to leave this out. But let me describe the scenario where I needed this in the past:

WPF controls show by default a red border that is defined via the Validation.ErrorTemplate. Now let's assume you have an INotifyDataErrorInfo implementation with a class that as FirstName property. It's required, so you return an error if the FirstName property is null or empty. Now you bound a TextBox to that FirstName property. When a user removes that firstname, a red border appears. Great, exactly what you want. But maybe you have at another place in the UI another control that is bound to the FirstName property too. For example a TextBlock in a Tab-Header. WPF will display on that TextBlock a red border too when there's a validation error. But this might not be what you want. You want the error just on the TextBox where the user can change the firstname, but not in the TextBlock in the Tab header. To get rid of the red border on the TextBlock, you don't have to edit the TextBlock's ErrorTemplate, instead you just turn of the INotifyDataErrorInfo validation on the TextBlock-FirstName-Binding by setting the ValidatesOnNotifyDataErrors property to false. That's the one and only use case I ever had for this property. :)

I hope this helps.

### Summary
Yes, I totally agree, having input validation just on x:Bind with support for INotifyDataErrorInfo is a good plan.

No, it's a fantastic plan, I'm already super excited!

**Comment by stevenbrix** [comment]
@thomasclaudiushuber thanks for the detailed follow up! The core scenarios you describe lineup with what we believed to be the real value driver, and so that's good to hear. The overall design of this feature and API has been in flux, mostly due to the previous work that WPF had. There were a few major things to call out here:

1. As discussed above, there were certain aspects of WPF's validation system that we didn't feel added any value, or were superseded by better patterns, and so we weren't planning on bringing them to UWP.

2. Certain features/functionality of WPF's validation system take advantages of core WPF features that don't exist in UWP. Most notably here is the adorner layer, which allows any UIElement in WPF to display validation visuals, not just Controls (via adding some template part).

3. UWP Xaml does not currently have the concept of attached events, and mirroring the Validation class would add the Validation.Error event. Not that is a deal breaker, just something to call out as we're generally wary of adding new concepts, as it only adds complexity. On top of that, since the work requires changes to the Control's Template, only the set of controls we provide the specific template changes for would work out of the box. Generally speaking, having an API that appears to work, but doesn't in some scenarios is not good practice, and so we were thinking it would be better to divorce ourselves from that model. This would mean having something else like a common interface and/or attributes that controls implement.

We're still in the process of understanding how to do API and spec reviews in the new and exciting open source world. We'd love to get the community's feedback, so hopefully we can get the API proposals out soon so the community can take a look and help us land something everyone is excited about!

**Comment by mdtauk** [comment]
@stevenbrix  When you do begin the open source spec for the validation features, will you provide the illustrations of how the validation will look, as well as the sample scenarios so the conversation can be a little broader for those of us more into the UI side, than the coding :)

**Comment by LucasHaines** [comment]
@mdtauk  Absolutely. The current UI we shared is still accurate. If you have feedback, feel free to provide on this thread.

**Comment by thomasclaudiushuber** [comment]
@LucasHaines The UI looks good to me. But I wonder how you can adjust how the errors are displayed. I haven't found anything regarding this and I don't know if you have already something to share in that area. Will there be something similar like the attached Validation.ErrorTemplate property from WPF?

**Comment by mrlacey** [comment]
Is there a timeline for when a formal spec/proposal will be released for this?
What can I do to help this get into the product sooner?

**Comment by knightmeister** [comment]
I'm with @mrlacey above... I could really use this now, so I'm happy to help out where I can.

**Comment by LucasHaines** [comment]
We are working on an open spec and I will let everyone know it's available in the spec repo.

**Comment by GeertvanHorrik** [comment]
Question: why not support IDataErrorInfo as well? It's supported in .net standard, and it would be weird not to support it? This should not affect performance for the people not using it since it could even be checked at compile time which interface to use for the binding?

**Comment by LucasHaines** [comment]
We started open spec review for Input Validation in the open spec repo. https://github.com/Microsoft/microsoft-ui-xaml-specs/pull/26

**Comment by knightmeister** [comment]
Hi guys, how are we tracking on this - do we have an ETA?

**Comment by LucasHaines** [comment]
@knightmeister We have an open spec created [here](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/26). Please feel free to participate in the review and help shape the feature. We are planning to ship this as part of WinUI 3.0. If you want more information on the WinUI 3.0 roadmap check out issue #717.

**Comment by LucasHaines** [comment]
Hey Everyone - please take a look at the latest samples in the spec. @stevenbrix made some great updates to the samples and address other feedback.

**Comment by weitzhandler** [comment]
This is a critical issue to me.
My entire XAML validation system is based on WPF validation rules, it reads the field and entity model from the `BindingExpression` and creates validation rules according to the DA validation attributes (see [this](http://bit.ly/2xV96ZL)).

**Comment by stevenbrix** [comment]
>  For example will it be possible for a container control (eg TabView or Pivot) to know that there is a child control in an invalid state ?

@tbolon yes, this is something that can be done, although it's unlikely there will be built in support in those containers for it. We've thought about creating a `Form` control that would have this functionality built-in, but it's probably on the backlog at the moment. Like in WPF, there is a `ValidationError` (that name could be wrong) event that each control fires when it becomes invalid/valid. This event isn't a routed event though, so you have to subscribe on each element that you are interested in listening to.

**Comment by stevenbrix** [comment]
> My entire XAML validation system is based on WPF validation rules,

@weitzhandler we currently aren't planning on adding something like the `ValidationRule` to WinUI. Is there something that using this gives you over using `ValidationAttributes` and having your model implement `INotifyDataErrorInfo`? I'd love to see what your Model and XAML look like to get a better understanding of your usage scenario.

**Comment by weitzhandler** [comment]
INDEI requires you to implement all your entities and sub entities.
Using rules, no changes are required, it all works automatically using validation attributes only.

**Comment by stevenbrix** [comment]
> INDEI requires you to implement all your entities and sub entities.

Can you elaborate more on this?

> Using rules, no changes are required, it all works automatically using validation attributes only.

For the most part, it all works automatically when using `ValidationAttributes` and `INotifyDataErrorInfo` as well. I'll show some code snippets here that are in the spec, just so we're (hopefully) on the same page.

FWIW, we're planning on having this `ValidationBase` class live in the Toolkit, so you don't have to write this boilerplate code yourself.

```csharp
public class ValidationBase : INotifyPropertyChanged, INotifyDataErrorInfo, IValidatable
{
   public event PropertyChangedEventHandler PropertyChanged;
   public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;

   protected void SetValue<T>(ref T currentValue, T newValue, [CallerMemberName] string propertyName = "")
   {
       if (!EqualityComparer<T>.Default.Equals(currentValue, newValue))
       {
           currentValue = newValue;
           OnPropertyChanged(propertyName, newValue);
       }
   }

   readonly Dictionary<string, List<ValidationResult>> _errors = new Dictionary<string, List<ValidationResult>>();
   public bool HasErrors
   {
       get
       {
           return _errors.Any();
       }
   }
   public IEnumerable GetErrors(string propertyName)
   {
       return _errors[propertyName];
   }

   private void OnPropertyChanged(string propertyName, object value)
   {
       PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
       Validate(propertyName, value);
   }

   private void AddErrors(string propertyName, IEnumerable<ValidationResult> results)
   {
       if (!_errors.TryGetValue(propertyName, out List<ValidationResult> errors))
       {
           errors = new List<ValidationResult>();
           _errors.Add(propertyName, errors);
       }

       errors.AddRange(results);
       ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
   }

   private void ClearErrors(string propertyName)
   {
       if (_errors.TryGetValue(propertyName, out List<ValidationResult> errors))
       {
           errors.Clear();
           ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
       }
   }

   public void Validate(string memberName, object value)
   {
       ClearErrors(memberName);
       List<ValidationResult> results = new List<ValidationResult>();
       bool result = Validator.TryValidateProperty(
            value,
            new ValidationContext(this, null, null)
            {
                MemberName = memberName
            },
            results
            );

        if (!result)
        {
            AddErrors(memberName, results);
        }
    }
}
```

Your model would then derive from that class and look like this:

```csharp
public class Person : ValidationBase
{
    private string name;
    [MinLength(4)]
    [MaxLength(6)]
    public string Name
    {	
        get { return name; }
        set { SetValue(ref name, value); }
    }
}
```

Assuming this `Person` class is set to a `ViewModel` property on your `Page`, you then bind to that, and validation happens automatically:

```xaml
<TextBox Text="{x:Bind ViewModel.Name, Mode=TwoWay}" />
```

**Comment by weitzhandler** [comment]
You're right, but that requires your entities to be non POCOs.
I should give it a try tho. Might not be so bad using an `EntityBase` class in a LoB app.

**Comment by rufw91** [comment]
**UWP doesn't have **HAVE** **Validation****!!!!!!! I just started a project. Back to WPF till a few years time.

**Comment by stevenbrix** [comment]
@rufw91 I know, right?!

What are your time constraints? The first implementation of this is in the WinUI3 alpha, and we should have a preview for WinUI desktop coming soon in the //Build timeframe. We plan on RTM'ing by the end of the year

**Comment by GeertvanHorrik** [comment]
> **UWP doesn't have **HAVE** **Validation****!!!!!!! I just started a project. Back to WPF till a few years time.

I've decided to stick with WPF for the upcoming years as well. I've done it all (WPF, SL, WP, UWP, etc), in the end only 1 tech seems solid, which is WPF. Maybe in a few years it could be interesting to check out where WinUI is, but I am tired of switching to new tech and being left alone in the dark. WPF is mature and works great. Maybe in a few years Windows as an OS isn't relevant at all, so let's await that before making any more investments in the platform.

**Comment by Nirmal4G** [comment]
`Windows.UI` is the UI stack used by Windows Shell teams. So, just for a minute forget us being customers and lets ask the shell teams... before the début of `Windows.UI` in Windows 8 publicly, Shell teams have been using `DirectUI` and it has similar features as it inspired WPF (If I'm not mistaken).

Don't they miss all the goodies that come with DirectUI and WPF Xaml?

Usually @microsoft responds well to internal customers (company teams) than to us. So, Why aren't **they** asking for it?

**Comment by weitzhandler** [comment]
> > INDEI requires you to implement all your entities and sub entities.
> 
> Can you elaborate more on this?

With my code, all POCO entities can remain free of implementing those interfaces, and remain without a base class. INPC is auto-implemented using Fody. Most I'd do is implement `IValidatableObject` when more refined validation is required. And it works on sub-types as well.

> FWIW, we're planning on having this `ValidationBase` class live in the Toolkit, so you don't have to write this boilerplate code yourself.

That's no good. What if my entities already have a base class? That's why I'm not such a big fan of `INotifyDataErrorInfo`. INPC properties are already a headache for themselves, especially when dealing with massive data apps with tons of entities.

**Comment by stevenbrix** [comment]
> It's quite shocking that UWP doesn't have proper validation yet. That's a crucial basic feature in almost any app and should be given first priority.

Yup, that's why we're working on it :)

> That's no good. What if my entities already have a base class? 

I'm just trying to understand your usage here. What is the base class, and could it derive from this class?

**Comment by weitzhandler** [comment]
> Yup, that's why we're working on it :)

Thank you, much appreciated. This, and the rest of your work.  
What does the label <kbd>needs-winui-3</kbd> imply?

> I'm just trying to understand your usage here. What is the base class, and could it derive from this class?

I'm using [OData Connected Service](https://bit.ly/3aw8A2A) to get entities generated on my client.
The generated objects are of this pattern:

```c#
public abstract partial class ContactBase : global::Microsoft.OData.Client.BaseEntityType, global::System.ComponentModel.INotifyPropertyChanged
```

**Comment by Felix-Dev** [comment]
> What does the label needs-winui-3 imply?

@weitzhandler 
It implies Input Validation will come with WinUI 3. (3.0 as currently planned). It probably won't come to WinUI 2 as that would require changes to the OS XAML code which is in maintenance mode by now.

**Comment by weitzhandler** [comment]
I'm using UWP in an Uno Platform app. Hopefully WinUI will be covered upon its release.
Thanks for the update tho!

**Comment by weltkante** [comment]
> I'm just trying to understand your usage here. What is the base class, and could it derive from this class?

Having a base class is often mandated by 3rd party, e.g. the persistence framework you use. UWP also demanding a base class just for validation is not an option. It needs to be an interface to be useful, none of our applications could make use of a base class.

**Comment by stevenbrix** [comment]
> Having a base class is often mandated by 3rd party, e.g. the persistence framework you use. UWP/WinUI also demanding a base class just for validation is not an option. It needs to be an interface to be useful, none of our applications could make use of a base class. 

Let's back up a bit. I think the last few comments I've made have caused a bit of confusion, which is my bad.

All that will be needed by the model is to implement `System.ComponentModel.INotifyDataErrorInfo` (or `Windows.UI.Xaml.Data.INotifyDataErrorInfo` for C++). The compiler will properly generate the code to hook that up with your view when using `{x:Bind}`.

The `ValidationBase` class I referred to above is a helper class that we were planning on shipping in the community toolkit that implements this, along with `INotifyPropertyChanged`. There's no need to derive from it if your use case doesn't allow it.

**Comment by weitzhandler** [comment]
What about validation attributes and [`IValidatableObject`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.ivalidatableobject?view=netcore-3.1), will this be respected by the UWP runtime?

How about other attributes, like the `MaxLength` attribute you mentioned earlier, will a `TextBox` max length be set automatically like in ASP.NET?
Same with [`DataTypeAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.datatypeattribute?view=netcore-3.1), [`EditableAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.editableattribute?view=netcore-3.1).
Also [`DisplayAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.displayattribute?view=netcore-3.1) for generating field headers.

**Comment by stevenbrix** [comment]
> What about validation attributes and IValidatableObject, are they gonna be respected by the UWP runtime?

So we currently are going to have support for the `Required` attribute. Some controls will automatically put a little asterisk next to the header text when this is used. We are open to supporting more in the future if you want. Please open a separate issue if you'd like to see more, and just to set expectations, any extra attribute support is highly unlikely to make the initial WinUI3 release.

As for `IValidatableObject`, I'm not sure how that would work. As a generalization, our engine works simply by listening to the `INotifyDataErrorInfo.ErrorsChanged` event. The actual validation of the model is up to the app developer, and generally done when transferring the value from the target to the source.

**Comment by weitzhandler** [comment]
Maybe as an alternative to the `ValidationBase` you proposed, there should be a set of extension methods, that run a validation process using [`Validator.TryValidateObject`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.validator.tryvalidateobject?view=netcore-3.1), and inject the validation results into the entity, notifying the changes?

**Comment by stevenbrix** [comment]
> Maybe as an alternative method to the ValidationBase you proposed, there should be a set of extension methods, that run a validation process using Validator.TryValidateObject, and inject the validation results into the entity, notifying the changes?

This is a great idea :)

**Comment by michael-hawker** [comment]
🦙 Wondering if the new source generator paradigms could help out here as well?

**Comment by thomasclaudiushuber** [comment]
Yes @michael-hawker, I like your thoughts!

 I had the same thoughts a while ago when I first read about the new source generator possibilities.

I think with the new source generator paradigms we can build / generate a lot of stuff behind the scenes for INotifyDataErrorInfo and INotifyPropertyChanged. 

I have a WPF course on Pluralsight that does exactly this with T4, INotifyPropertyChanged and INotifyDataErrorInfo, including Data Annotations for validation: https://www.pluralsight.com/courses/wpf-mvvm-advanced-model-treatment
The course also uses custom Attached Properties to hook up things. 
The goal of this course is to show how you can build a MVVM WPF app that shows on every input field if it was changed and what was the original value. 

So, definitely, generating things and maybe ending up with a library (maybe as part of WCT) would be awesome. And I'd love to redo that course for WinUI. :)

**Comment by weitzhandler** [comment]
Ideally, WinUI should automatically inject all necessary UI according to the bound model's DA attributes, not limited to validation but also display name, max-length, data-type and so on.
Just like in ASP.NET Core (see [this](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-3.1#client-side-validation).

**Comment by rufw91** [comment]
> @rufw91 I know, right?!
> 
> What are your time constraints? The first implementation of this is in the WinUI3 alpha, and we should have a preview for WinUI desktop coming soon in the //Build timeframe. We plan on RTM'ing by the end of the year

I hope so, i had to switch back to WPF, im actually just about done now.

**Comment by weitzhandler** [comment]
I opened [this](https://bit.ly/3gcMJ3a) one.

**Comment by Sergio0694** [comment]
To follow up on this, after a conversation with @stevenbrix on Twitter, I've integrated the base `INotifyDataErrorInfo` implementation he provided [here](https://github.com/Microsoft/microsoft-ui-xaml/issues/179#issuecomment-601296844) into a new `ObservableValidator` class being included in the new MVVM Toolkit (the `Microsoft.Toolkit.Mvvm` library), which also implements `INotifyPropertyChanged` and `INotifyPropertyChanging` 😄

You can see the related code and a recap of everything related to that library [in this issue](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3428).
I've also included links there to the original issue for the MVVM Toolkit, as well as more docs on it for new devs.

**Comment by worldbeater** [comment]
Could you please give us an update on this? Curious if UWP 10.0 18362 supports `INotifyDataErrorInfo`. According to the following documentation page, https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo?view=netcore-3.1 `INotifyDataErrorInfo` "applies" to UWP 10.0. However, with UWP 10.0 18362 installed and targeted, the text boxes aren't displaying `INotifyDataErrorInfo` validations. The text boxes are described in XAML as follows:
```xaml
<TextBox Text="{x:Bind ViewModel.Username, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
```
And we use [`ReactiveValidationObject`](https://github.com/reactiveui/ReactiveUI.Validation/blob/main/src/ReactiveUI.Validation/Helpers/ReactiveValidationObject.cs) as `INotifyDataErrorInfo` implementation here.

**Comment by stevenbrix** [comment]
@worldbeater this is only in available in the latest previews of WinUI3, are you using that?

**Comment by worldbeater** [comment]
Thank you @stevenbrix! Going to try out WinUI3. Glad to hear `INotifyDataErrorInfo` and UWP are befriending, finally ✨

**Comment by stevenbrix** [comment]
@worldbeater, it's still in preview, so we would love to get your feedback on it! 💪

**Comment by maxkatz6** [comment]
@worldbeater by the way, docs says it's available on UWP 10.0, but it doesn't mean it's supported by it. Documentation isn't clear enough.

**Comment by rezss** [comment]
Hey @stevenbrix, don't know where this issue fits best, but your implementation you kindly [provide here](https://github.com/microsoft/microsoft-ui-xaml/issues/179#issuecomment-601296844) has a bug. As soon as a property failed to validate and is added to the _errors list, HasError will always return true, as you never remove the empty ValidationResult list from the errors Dictionary itself. Interestingly enough, though, that's the only implementation that will clear the actual error from the GUI. The reason is, in the autogenerated validation code, there is the following guard:

```
private void UpdateErrors_(global::Microsoft.UI.Xaml.Controls.Control control, global::System.ComponentModel.INotifyDataErrorInfo sender, string propertyName)
{
    if (sender.HasErrors)
    {
        UpdateInputValidationErrors(control, sender.GetErrors(propertyName));
    }
}
```

preventing the wipe of the inputValidationErrors if INotifyDataErrorInfo does not return true on HasErrors...

**Comment by knightmeister** [comment]
Hi @stevenbrix, I see in the [roadmap](https://github.com/Microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md) that this is now labelled a stretch goal.

I stopped working in Win UI because it felt too immature, but I really want to come back because I love the framework. Validation is absolutely critical and my concern is it's going to keep slipping (this has been coming soon for ages, iirc).

When is it likely that we're going to have this ability? You cannot develop robust LOB without competent validation handling, and throwing up a MsgBox (`ContentDialog`) like it's 1990 just doesn't cut it for me.

**Comment by sonnemaf** [comment]
I totally agree with @knightmeister. Input validation is a MUST HAVE. Without it it is totally useless for LOB apps. How can I sell WinUI to WPF developer's without it? It must be reconsidered.

**Comment by weitzhandler** [comment]
Totally. Input validation is crucial.
Btw here's my related issue about generalvalidation in .NET:
[https://github.com/dotnet/extensions/issues/3379#winui](https://bit.ly/3gcMJ3a).

**Comment by GeertvanHorrik** [comment]
> How can I sell WinUI to WPF developer's without it?

Maybe you shouldn't? I think at this moment (not sure it will ever change in the future), the most reasonable path for most LOB developers is sticking with WPF, then consider Blazor / any other web technology standard next.

**Comment by knightmeister** [comment]
@stevenbrix are we able to get an official position from Microsoft on this please - when/if it is likely to happen. I'm looking at starting a new project soon and validation is a major consideration for it as a LOB application.

**Comment by Noemata** [comment]
@GeertvanHorrik , WPF was fantastic in its day and is still viable.  Suggesting it as an alternative to UWP for those of us that were coaxed away from WPF by Microsoft's UWP dream feels like a stab in the heart after the stab in the back with Silverlight.

On this particular topic, promises were made in 2018: https://www.youtube.com/watch?v=4chzzzAZRNQ

There was all sorts of hoopla about how great UWP was going to be for LOB.  Telerik's UWP lib was a stopgap for some of the missing pieces.  The DataGrid took years, not that there was any good reason for the delay given the code was all there in Silverlight (I watched the Silverlight->UWP porting process on Github transpire, painful).  The whole LOB proposition took a giant step backwards when Siverlight was replaced by UWP.  Now we have the WinUI replacement which is yet another step backwards relative to where UWP is now.  Microsoft, it will continue to be unpleasant to visit these issue entries until promises start being fulfilled in a big way.

No one expected UWP to be all there on day one.  By year 3 or 4 patience ran out.  WinUI 3.x has no runway left over to leverage.  The griping / sour grapes / whining won't go away given the history at hand.  The fix is to start delivering on promises made and fast.

**Comment by knightmeister** [comment]
>No one expected UWP to be all there on day one. By year 3 or 4 patience ran out. WinUI 3.x has no runway left over to leverage. The griping / sour grapes / whining won't go away given the history at hand. The fix is to start delivering on promises made and fast.

I completely agree. It's incredible to me that a company as large as Microsoft can't even get validation - which was implemented in WPF years ago (they're using the same pattern!) - working in WinUI. Validation is absolutely critical for any LOB scenario. The silence is absolutely deafening on this one, there has been no response by @stevenbrix with me questioning whether this is going to be supported in WinUI3. 

Fundamentally I'm getting to a point that I need to determine the architecture for something that is planned to have a long life, and WinUI is pretty much off that list now because we need support for complex validation, and looks like I will need to use WPF for this project.

**Comment by mrlacey** [comment]
The title of this issue specifically mentions UWP but is it reasonable to assume this issue is also being used to track the implementation of Validation support for WinUI for Desktop too?

If so, can it also be assumed (based on this being in the freezer) that there's won't be validation support in version ReunionV0.5?

When might validation support be available? reunion-v0.8? reunion-v1.0?  sometime in 2022?

**Comment by thomasclaudiushuber** [comment]
Hey @mrlacey, thanks, I've adjusted the title of this issue. Input validation is mentioned in the table in the [roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md). It's a stretch goal for WinUI 3.0 release, and planned/included with post-3.0-releases.

**Comment by michael-hawker** [comment]
@MikeHillberg should this be re-assigned to you now? Who's the right owner for this now?

FYI getting meaningful feedback for this feature in the WinUI 3 Previews is blocked by issue #4092. It should be a simple fix, so it'd be nice to get it in so that folks can test it out.

**Comment by michael-hawker** [comment]
@ryandemopoulos is there plans to service #4092 in the near future so this feature can get feedback sooner than later for the future? Who should be assigned to his issue now as the PM?

**Comment by knightmeister** [comment]
Guys, can we get an update please? Is this still likely to make WinUI 3?

**Comment by thomasclaudiushuber** [comment]
Hey @knightmeister, that's a great question.

Let's sum up the state. Input validation is part of the Preview releases (0.5 and 0.8), but not of the stable releases.

Docs to Project Reunion 0.5 stable release (https://docs.microsoft.com/en-us/windows/apps/winui/winui3/ <- note that this link points to the latest stable release. So depending when you read this, it might not point to 0.5 anymore) say this:
> As this is a stable release, preview features have been removed from this version of WinUI 3. You can still access these features by using the current preview version of WinUI 3. Please note the following key features are still in preview, and work to stabilize them is ongoing:
...
Input validation

You find Input validation in the Project Reunion 0.5 preview and also in 0.8 preview (see also [0.8 release notes](https://docs.microsoft.com/en-us/windows/apps/winui/winui3/release-notes/release-notes-08-preview)).

According to the [WinUI roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md) input validation does not come to the Project Reunion 0.8 stable release that is planned for June and that is ready to be used for production apps. But input validation will come in a future stable release. It's not clear from the roadmap if input validation is coming to Project Reunion 1.0 or not. I think once the table gets updated to include the 1.0 release, we know more. 

So, right now, I guess the team is busy with the 0.8 stable release, and then we might see the planning for the 1.0 release.

**Comment by knightmeister** [comment]
Thanks for the quick reply @thomasclaudiushuber. I've seen the roadmap, I'm just hoping that it's not going to slip and will still make it in the near future.

I need this to sell the business on migrating our stack to WinUI - because we make extensive use of this pattern in WPF and it works well.

**Comment by weitzhandler** [comment]
Yes thank you both. I'd like to share I have similar standing point where input valuation is probably the most compelling feature missing.

**Comment by rufw91** [comment]
I cant believe validation is still not yet complete!!!!! This is some kind of a bad joke. I'm sticking with WPF until they sort out the mess that is UWP/WinUI/MAUI/Project Reunion or whatever it is that they are calling the desktop UI stack.

**Comment by sonnemaf** [comment]
The new Windows 11 input controls (TextBox, NumberBox, ComboBox, etc) get a new look. They will get an accent color bottom border. They look similar to the Google [Material Design](https://material.io/components/text-fields#usage) ones.

Will this bottom border also be used for validation? Red when invalid? 

![Example](https://docs.microsoft.com/en-us/windows/apps/get-started/images/great-apps/controls.png)

**Comment by mdtauk** [comment]
Will need more than just a colour change, especially if users have a red accent colour.

Will need to display an icon and validation message to be accessible and clearly define an error state.

**Comment by sonnemaf** [comment]
@mdtauk I agree that only a color changes is not enough. I just posted my reply to get the discussion going. 

Not sure if we need an icon. If so maybe the icon should be placed inside the element. Not next to it. This will make the alignment of multiple valid and invalid controls easier/better.

**Comment by cesarchefinho** [comment]
Please 

make possible to stop validation in xaml and enable it again in viewmodel.

actualy we have clearerrors in viewmodel, but there are same use cases that we want to prevent that binding validade before an RealyCommand fired from a button in view. in that case only when the button fired the relaycommand we want to execute validation in the RelayCommand using a function like ValidateAll ().

maybe can be an xaml attribute or atached property of controls with onclick events, like buttons, hyperlinkbuttons, etc NotGenerateValidation 

this is the same usecase in web development like html below

<input type="submit" onclick="this.form.setAttribute('novalidate', 'novalidate'); .......

**Comment by Sergio0694** [comment]
> "there are same use cases that we want to prevent that binding validade before an RealyCommand fired from a button in view"

Unless I'm misinterpreting your message, this is already possible today, if you just skip the validation in the property setters?
That is, assuming you're using `ObservableValidator` from the MVVM Toolkit (given you mentioned the `ValidateAll` method):

```csharp
private string? name;

public string? Name
{
    get => name;
    set => SetProperty(ref name, value); // Not passing "true" here skips validation
}

[ICommand]
private void Submit()
{
    ValidateAll(); // Only validate properties when submitting

    // Check for errors, rest of the logic...
}
```

This will cause all validatable properties to not make any errors show up, until the button is pressed, which will then validate all existing properties. If you instead meant that you only wanted individual properties to _start_ validating only after the button had been pressed at least once, then can't that still be done in your viewmodel by eg. setting some local field to indicate that the buton has been pressed, and then passing that to the `bool` flag to `SetProperty` calls in each property setter?

Either way I think I'm not following how this request would be tied to WinUI itself 🤔

**Comment by Brouilles** [comment]
It is also coming in UWP or only Windows App SDK?

**Comment by thomasclaudiushuber** [comment]
Hey @knightmeister, @stevenbrix switched teams at Microsoft and is no more in the WinUI team. (Steve, if you read this: Hope you're doing well :-))

The input validation feature is in the table that you find in the roadmap: https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md 

According to the roadmap, input validation is an experimental feature of Windows App SDK 1.0 in Q4 2021, and it will be a supported feature in a future release. So, I guess there is no concrete timeframe yet for this feature.

> It would be great for it to get some visibility as it is absolutely crucial for LOB applications.

Absolutely!

**Comment by knightmeister** [comment]
Hi @thomasclaudiushuber I've finally had a chance to revisit this after nagging about it for ages.

Is there any sample code or documentation available for this? Or should it just work in the same way as it does in WPF?

Cheers.

**Comment by thomasclaudiushuber** [comment]
@knightmeister Good questions, I actually also don't know who is currently working on this and what the current state is. (@marb2000 might know)

What do we have right now?

1. The PR for the specs is still not merged. But I guess right now it is our best (and only) doc that we have: https://github.com/microsoft/microsoft-ui-xaml-specs/pull/26/files?short_path=990dcdf#diff-990dcdfa22e86b6de4de494a56f0a9347135e3e24529939b057678881064a141
2. The roadmap of WinUI 3 says that it's an experimental feature in Windows App SDK 1.0, so I would assume that it should be included in the experimental Windows App SDK 1.0 channel here https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/experimental-channel, but I haven't tried that channel yet since 1.0 was released.

**Comment by rufw91** [comment]
I tried the code provided. It simply doesn’t work. I wouldn’t imagine implementing this is so difficult as to take this long. The could copy paste some code from WPF repo :)
________________________________

**Comment by mrlacey** [comment]
> I tried the code provided. It simply doesn’t work. I wouldn’t imagine implementing this is so difficult as to take this long. The could copy paste some code from WPF repo :)

I believe a big challenge of implementing this is that the desired implementations require that the underlying controls (`TextBox`, etc.) need a level of capability to display error states and indicators and as they were originally created as part of Windows 10 these weren't included. Now that the XAML stack has been separated from the OS, it makes it possible to add these in for the future (itself not a small task) but also there's the need to polyfill for the versions of the controls on all the older supported versions of the OS.

**Comment by knightmeister** [comment]
>I believe a big challenge of implementing this is that the desired implementations require that the underlying controls (TextBox, etc.)

I suspect you're right. This whole issue is pretty frustrating though, validation is absolutely crucial to even the most trivial LOB application. I want to sell WinUI to the business and I'm pretty passionate about this project but I can't do it without validation :(

**Comment by thomasclaudiushuber** [comment]
> validation is absolutely crucial to even the most trivial LOB application. I want to sell WinUI to the business and I'm pretty passionate about this project but I can't do it without validation :(

I fully agree to this @knightmeister. Missing input validation is a **show stopper** for WinUI in many companies. I mean, you could validate user input on your own and show some TextBlocks here and there in the UI. But usually the framework is not considered to be made for LOB apps without this feature.

I think it would be amazing if we could use the same INotifyDataErrorInfo ViewModels in WPF and WinUI. The more I think about mixing WPF and WinUI with XAML Islands, the more sense it makes to me to have the same validation stack in the ViewModels.

**Comment by thomasclaudiushuber** [comment]
> I wouldn’t imagine implementing this is so difficult as to take this long. The could copy paste some code from WPF repo :)

@rufw91 I think it's not as easy. WinUI is a native stack that can be used with C++, WPF on the other side is mostly managed code (and so also the input validation implementation). In addition, input validation in WinUI should work with `x:Bind`, which means it's a complete different challenge than with the `Binding` markup extension in WPF. Only from a user/dev perspective, it can be very similar.

Note that the Windows App SDK has now a ProductBoard, and seems Input Validation is there in the backlog:
https://portal.productboard.com/winappsdk/1-windows-app-sdk/tabs/7-backlog

**Comment by knightmeister** [comment]
>I fully agree to this @knightmeister. Missing input validation is a show stopper for WinUI in many companies. I mean, you could validate user input on your own and show some TextBlocks here and there in the UI. But usually the framework is not considered to be made for LOB apps without this feature.

Exactly. We have a properly factored WPF application that makes extensive use of `INotifyDataErrorInfo` so we're hoping to do a lift and shift of view models and get WinUI for the cost of a new UI.

There will of course need to be some changes (particularly around the windowing space) but for now our main stopper is this.

**Comment by listepo** [comment]
@Redth will be there something similar for the MAUI?

**Comment by Noemata** [comment]
@thomasclaudiushuber , validation was promised for UWP in 2018.  There were even some demos.  It's a little late to be discovering this is hard.  What's interesting, now that UWP has been dumped, CSWinRT is proving to be hard too.  Performance challenges are mounting.  The only viable path for WinUI at the moment appears to be good old C++.  Very, very strange trajectory.  Managed WinUI is turning into yet another replay of Silverlight.  Looks like anything with a Silverlight lineage is doomed, at least for the managed side of things, is doomed.  (WPFe, Silverlight, Silverlight on Windows Phone, UWP in Win8, UWP in Win10, UWP on HoloLens, UWP on Surface Hub, UWP on IoT ... all dead).

**Comment by thomasclaudiushuber** [comment]
@listepo I'd recommend to open an issue in the .NET MAUI repo, then you can link from there to this issue here. This will give you higher chances to get a response from @Redth and the .NET MAUI team. This is a native WinUI issue that is not related to .NET MAUI, so it's a bit off-topic here. Xamarin.Forms had input validation support, but AFAIK no direct support for `INotifyDataErrorInfo`. So, let's see what will happen there, and let's see what they do to bring something to the created WinUI target. (When you create an issue in the .NET MAUI repo (https://github.com/dotnet/maui) you can add a link there, then we'll see it also here.

@Noemata It's still part of the roadmap and moved to an after 1.1 release: https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md So, I expect that we'll see something there in the future.

**Comment by luisgarciau** [comment]
> > I wouldn’t imagine implementing this is so difficult as to take this long. The could copy paste some code from WPF repo :)
> 
> @rufw91 I think it's not as easy. WinUI is a native stack that can be used with C++, WPF on the other side is mostly managed code (and so also the input validation implementation). In addition, input validation in WinUI should work with `x:Bind`, which means it's a complete different challenge than with the `Binding` markup extension in WPF. Only from a user/dev perspective, it can be very similar.
> 
> Note that the Windows App SDK has now a ProductBoard, and seems Input Validation is there in the backlog: https://portal.productboard.com/winappsdk/1-windows-app-sdk/tabs/7-backlog

@thomasclaudiushuber working on a WinUI application and have no input validation implemented  :) . I went to the Windows App SDK productboard you mentioned and voted this item as critical. Hopefully this will help this feature to be implemented. Thank you for staying on top of this!

**Comment by BusterIT** [comment]
Hello together. I was also wondering about that this basic functionality has not been implemented, yet. After I read the discussion on this thread, I decided to create an own easy solution to provide UI validation which is very similar to Blazor-Form-Validation. It supports x:Bind & DataAnnotations validation attributes and is easy to implement. Including submit button (disabled on invalid model state) and validation of nested objects. 

Implementation example:
```
     
... xmlns:validation="using:WinUIValidation" ...

    <validation:Form Model="{x:Bind Person}" OnValidSubmit="{x:Bind Person.SubmitCommand}">
        <StackPanel HorizontalAlignment="Center" VerticalAlignment="Center" MaxWidth="300">
            
            <!--Adds red border to content-->
            <validation:ValidationFor PropertyName="Name">
                <TextBox Text="{x:Bind Person.Name, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
            </validation:ValidationFor>
            
            <!--Just shows validation message-->
            <TextBox Text="{x:Bind Person.Address.CityName, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
            <validation:ValidationFor PropertyName="Address.CityName" />

            <validation:SubmitButton Content="Submit" HorizontalAlignment="Stretch" Margin="0,15,0,0"/>
        </StackPanel>
    </validation:Form>

```
![image](https://user-images.githubusercontent.com/70801990/168973516-b749cdd8-ad7f-4a5f-91f0-65ebac13037f.png)

- no additional code behind is required, just XAML
- no implementation of IDataErrorInfo is required, just INotifyPropertyChanged

You can get the package from nuget: https://www.nuget.org/packages/WinUIValidation
or have a look to repository (including example project) here: https://github.com/BusterIT/WinUIValidation

**Comment by BusterIT** [comment]
> I'm using this validation library with WinUI3 and it is working great.
> 
> The only issue is, when I'm using NavigationCacheMode="Enabled" in XAML to cache the data entered. When I comment <validation:Form Model="{x:Bind pkg}" OnValidSubmit="{x:Bind pkg.SampleCommand}" ... tags, it Cache working fine. If I uncomment <validation:Form Model="{x:Bind pkg}" OnValidSubmit="{x:Bind pkg.SampleCommand}" it throws exception "no installed components were detected. (0x800F1000) in winui3".
> 
> Any help in this is much appreciated.

Hi, could you post your example code, both the one that works and the one that throws the exception?
At what point does it occur? When you navigate between pages? And when using NavigationCacheMode="Disabled" there are no problems?

**Comment by BusterIT** [comment]
> I'm using this validation library with WinUI3 and it is working great.
> 
> The only issue is, when I'm using NavigationCacheMode="Enabled" in XAML to cache the data entered. When I comment <validation:Form Model="{x:Bind pkg}" OnValidSubmit="{x:Bind pkg.SampleCommand}" ... tags, it Cache working fine. If I uncomment <validation:Form Model="{x:Bind pkg}" OnValidSubmit="{x:Bind pkg.SampleCommand}" it throws exception "no installed components were detected. (0x800F1000) in winui3".
> 
> Any help in this is much appreciated.

@Balu5a3 - Thanks for your feedback. I fixed this issue now.
@thomasclaudiushuber - Maybe one day in distance it´s ready to be used in enterprise solutions, too ;)

**Comment by cesarchefinho** [comment]
I maded an error validation on winui 3

It uses description property of controls that have it, the mvvm model error property and binding mechanism. 

For controls that don't have description property (for example, a checkbox) you can bind error message to another control, like an textblock  bellow the validated control. 

Also these extensions simplify binding from data annotation of viewmodel and reduces the bloated markup of xaml

Can you look to this implementation? 

https://github.com/cesarchefinho/MVVMBindingExtensions

**Comment by sude22** [comment]
We are at the end of 2022, and still there is no real progress with this issue?
Like other people said before, for LOB apps a simple validation via INotifyDataErrorInfo is a minimum requirement.
Or is WinUI3 only designed to build small example apps?

**Comment by omikhailov** [comment]
I still think that INotifyDataErrorInfo is not more than a dusty garbage. No other framework puts all errors into one heap the way WPF does.

If someone needs validation, then in #8011 I have given links to my library, which has a MIT license, so that it can be modified as you like. Although, of course, it would be better if something similar was in the Community Toolkit.

**Comment by r7dev** [comment]
With Visual Studio Lightswitch we had a great data validation framework. They could invite the old architects to participate in this project.
[Article about the framework.](https://www.codemag.com/article/1105091/Understanding-the-Data-Validation-Framework-in-Visual-Studio-LightSwitch)

**Comment by jrovny** [comment]
Hi, all. Wanted to follow up on this and see if this still on the roadmap or what the current status is. Thanks!

**Comment by thomasclaudiushuber** [comment]
Hi @jrovny, yes, it is still on the roadmap. [Here on the WinUI roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md) you find a feature table that contains a record called **Input validation for data fields**. But there's no communication when this experimental feature will be available as a supported feature.

**Comment by ghost1372** [comment]
Input validation API added in v.1.5 exp2😁

**Comment by sude22** [comment]
I can't see it in the ReleaseNotes....

**Comment by ghost1372** [comment]
> I can't see it in the ReleaseNotes....

Many changes are not mentioned
Also there is some new controls like PagerControl

See full changes here
https://omds.xaml.dev/WinAppSDKv1.4_v1.5e2.html

**Comment by MartinRobins** [comment]
> Input validation API added in v.1.5 exp2😁

I cannot see anything that indicates how to use this API; are there any examples?

**Comment by ghost1372** [comment]
> > Input validation API added in v.1.5 exp2😁
> 
> I cannot see anything that indicates how to use this API; are there any examples?

i did not test, but you can try winui 2 examples, mybe you need to changes somethings...
https://github.com/microsoft/WinUI-Gallery/blob/winui2/WinUIGallery/ControlPages/InputValidationPage.xaml

**Comment by MartinRobins** [comment]
> > > Input validation API added in v.1.5 exp2😁
> > 
> > 
> > I cannot see anything that indicates how to use this API; are there any examples?
> 
> i did not test, but you can try winui 2 examples, mybe you need to changes somethings... https://github.com/microsoft/WinUI-Gallery/blob/winui2/WinUIGallery/ControlPages/InputValidationPage.xaml

Gave it a try, cannot seem to get anything out of it.
Also, example page seems to be generating error messages in the page code rather than just relying on INotifyDataErrorInfo?

**Comment by thomasclaudiushuber** [comment]
Hey everyone, the InputValidationPage.xaml file in the GalleryApp is an old thing. If you do a git blame on it, you'll see that it has been around since 2018, and last changes were made in 2020 (4 years ago).

Here's the blame on the file: https://github.com/microsoft/WinUI-Gallery/blame/main/WinUIGallery/ControlPages/InputValidationPage.xaml

 So, it might not be an up-to-date version of the implementation that we might see in the future. :) 
 
 Let's keep an eye on it. If I'll find something out, you can be sure that I'll follow up with all of you here.

**Comment by MartinRobins** [comment]
Given that this was in the experimental build(s), does anybody know if it is also in the pre-release build?
(even though I still cannot work out how to use it in the experimental)

**Comment by ghost1372** [comment]
> Given that this was in the experimental build(s), does anybody know if it is also in the pre-release build? (even though I still cannot work out how to use it in the experimental)

i checked preview1 and Currently not available it seems that preview1 contains exp1 stuff, so we should wait for preview2 or stable release

**Comment by SoggyBottomBoy** [comment]
This doesn't appear to be part of the road map for 1.6. Is this true? I'm surprised this is not one of the most requested features. Can anyone provide an update on this?

**Comment by cesarchefinho** [comment]
No one more needs inputvalidation and no one more need winui3

Avalonia has inotifydataerrorinfo implemented and it works!

Also avalonia has an xaml designer and it works! 

Also avalonia works with visual studio, visual studio code and with jet brains rider (especialy on mac) 

Also avalonia compiles to windows, macos, linux, androi, ios, tizen and even to browser!

Winui is dead. Maui is dead at born... long live to avalonia.

**Comment by r7dev** [comment]
Why am I not surprised by this?
[Removed input validation page](https://github.com/microsoft/WinUI-Gallery/commit/2538298dcb169c4c1264dc07eb9490552ec98c48#diff-a917b987d4d499462be12ea4c080002c7c2317926b57b942004ee8181bce19e6) from the WinUI Gallery app.

**Comment by CastelloBrancoTecnologia** [comment]
> Why am I not surprised by this? [Removed input validation page](https://github.com/microsoft/WinUI-Gallery/commit/2538298dcb169c4c1264dc07eb9490552ec98c48#diff-a917b987d4d499462be12ea4c080002c7c2317926b57b942004ee8181bce19e6) from the WinUI Gallery app.

Simple! You use avalonia ui that have inputvalidation and plus an visual studio xaml designer. No one more needs winui

Microsoft put winui in tashcan, it is simple.

**Comment by soroshsabz** [comment]
ITNOA

Any update?
